tests/end-to-end.rs

```diff
@@ -22,86 +41,366 @@ macro_rules! test_case {
+            println!("Copying {filename} into {test_working_dir}");
+            copy(entry.path(), format!("{test_working_dir}/{filename}")).unwrap();
+        }
+    }
+
+    let cdk_app_file_path = format!("{expected_outputs_dir}/{cdk_app_filename}");
+    println!("Checking for cdk app file in snapshots: {cdk_app_file_path}",);
 
-test_case!(documentdb, "DocumentDbStack");
+    // If the app file already exists in the snapshot, copy to the test's working dir
+    let app_dst_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+        .join(test_working_dir)
+        .join(cdk_app_filename);
+    println!("DEBUG: app_dst_path: {:?}", app_dst_path);
+    if let Ok(mut cdk_app_file) = snapshots_zip.by_name(&cdk_app_file_path) {
+        let mut contents = Vec::<u8>::new();
+        cdk_app_file.read_to_end(&mut contents).unwrap();
+        println!("An app file already exists. Copying it to: {:?}", app_dst_path);
+        let mut file = File::create(app_dst_path).unwrap();
+        file.write_all(contents.as_bytes()).unwrap();
 
-struct UpdateSnapshot<'a> {
-    path: &'static str,
-    actual: &'a str,
-    expected: &'a str,
+    } else if std::env::var_os("UPDATE_SNAPSHOTS").is_some() {
+        // If it does not already exist, create new app file
+        println!("UPDATE_SNAPSHOTS=true, and there is no existing app file, creating default one at {}...", app_dst_path.to_str().unwrap());
+        let mut file = File::create(app_dst_path).unwrap();
+        let code: CodeBuffer = CodeBuffer::default();
+        // Recommend to manually instantiate a stack for each combination of parameters/conditionals that should be tested
+        code.line("// auto-generated! a human should update this!");
+        code.line("import * as cdk from \"aws-cdk-lib\";");
+        code.line(format!(
+            "import {{ {} }} from \"./stack\";",
+            cdk_stack_classname
+        ));
+        let app = code.indent_with_options(IndentOptions {
+            indent: INDENT,
+            leading: Some("const app = new cdk.App({".into()),
+            trailing: Some("});".into()),
+            trailing_newline: true,
+        });
+        let app_props = app.indent_with_options(IndentOptions {
+            indent: INDENT,
+            leading: Some("defaultStackSynthesizer: new cdk.DefaultStackSynthesizer({".into()),
+            trailing: Some("}),".into()),
+            trailing_newline: true,
+        });
+        app_props.line("generateBootstrapVersionRule: false,");
+        code.line(format!("new {}(app, \"Stack\");", cdk_stack_classname));
+        code.line("app.synth();");
+        code.write(&mut file).unwrap();
+    } else {
+        // Fail the test to prevent tests without snapshots from succeeding in CI/CD
+        panic!("CDK App file not found at {}. If you are developing a new test, set UPDATE_SNAPSHOTS=true in your environment variables and the test will create a default app file.", cdk_app_filename);
```

Comment:
This will make the test fail with a diff between a default app and whatever was synthesized. Would it be preferable to fail with a diff between an empty file and the code that was synthesized?

Reply:
The `App.ts` file is not part of the snapshot diffing stuff. It is just there so that we have a complete app to synth. And so we can test all combinations of parameters to the stack. 

I don't really understand your question. Can you clarify?

Reply:
ah I thought this was part of the snapshot diff, as in the base expected template. I'm not really sure why we need this to have a complete app to synth, could you elaborate a bit on why we need empty CDK apps on disk?